lib/AST/NameLookup.cpp

```diff
@@ -137,14 +137,40 @@ void DebuggerClient::anchor() {}
 void AccessFilteringDeclConsumer::foundDecl(
     ValueDecl *D, DeclVisibilityKind reason,
     DynamicLookupInfo dynamicLookupInfo) {
-  if (D->hasInterfaceType() && D->isInvalid())
-    return;
   if (!D->isAccessibleFrom(DC))
     return;
 
   ChainedConsumer.foundDecl(D, reason, dynamicLookupInfo);
 }
 
+void UsableFilteringDeclConsumer::foundDecl(ValueDecl *D,
+    DeclVisibilityKind reason, DynamicLookupInfo dynamicLookupInfo) {
+  // Use within decl's own initializer
+  if (auto *VD = dyn_cast<VarDecl>(D)) {
+    if (auto *init = VD->getParentInitializer())
+      if (SM.rangeContainsTokenLoc(init->getSourceRange(), Loc))
+        return;
+  }
+
+  switch (reason) {
+  case DeclVisibilityKind::LocalVariable:
+    // Use of 'TypeDecl's before declaration is allowed
+    if (!isa<TypeDecl>(D) && !SM.isBeforeInBuffer(D->getLoc(), Loc))
+      return;
+    break;
+  case DeclVisibilityKind::VisibleAtTopLevel:
+    // TODO: Implement forward reference rule for script mode? Currently,
+    // it's not needed because the rest of the file hasn't been parsed.
+    // See: https://bugs.swift.org/browse/SR-284 for the rule.
```

Comment:
I didn't have time to look into this one, but is this actually true? I know they're not in the results since the tests pass, which check this case... so I assume there's something somewhere?

Reply:
"hasn't been parsed" is not correct. We *skip* the rest of the file.
https://github.com/apple/swift/blob/a69502e6b1fc331cec4033d54723807d6c19b3e9/lib/Parse/ParseStmt.cpp#L248-L251
https://github.com/apple/swift/blob/a69502e6b1fc331cec4033d54723807d6c19b3e9/lib/Parse/ParseDecl.cpp#L3797-L3802

Reply:
Ah cool, thanks. I assume this is something that will stick around - should I just remove that comment? (it was in the original class)

Reply:
I think leaving `TODO` is OK for now. It'd be great if we actually implement it some day.

Reply:
I just removed the FIXME and re-worded the default case a little.